server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/BalancedShardsAllocator.java

```diff
@@ -159,6 +162,13 @@ public void allocate(RoutingAllocation allocation) {
         balancer.allocateUnassigned();
         balancer.moveShards();
         balancer.balance();
+
+        nodeWeightStatsSupplier = new BalancedShardsAllocatorNodeWeightStats(balancer, weightFunction)::get;
```

Comment:
Apologies I am going to repeat my comment from the other place to try promote it one more time 🙏  I think we should return this statsSupplier instead of keeping it around as a field which unnecessarily lingers around and retains references to a bunch of objects after the computation is done. It is likely trivial in practice since this method should be called often enough so that it does _not_ stay around for too long. But conceptually still feels a bit odd. The method computes something that is immediately needed (when necessary) by the caller. To me, this feels naturally as returning a value so that the order is clear, i.e. you cannot use the value before calling this method. I don't believe we'll do that in the code. But keeping it as a field opens up for that possibility. Personally I also generally prefer fewer mutable fields.

Reply:
I am a bit concerned about this one. It captures a reference to the balancer and keeps it in memory for some time.
Balancer might be a somewhat sizeable (it has its internal model representing every node, every index and every shard).

May be it is worth to eagerly compute stats and keep them in the field? Per node data structure sounds much more compact.
Possibly there could be another way to reduce the scope of the stats

Reply:
If we return and assign it to a local variable in a tight loop. It and its referenced balancer should not be kept around for long after each loop. My suggestion was assuming we don't want to eagerly compute the stats. If that is not the case, I'd suggest we return the stats from this method.

Reply:
> May be it is worth to eagerly compute stats and keep them in the field? Per node data structure sounds much more compact.

Isn't that what I had first (returned via `RoutingNodes`) and you suggested it might be expensive? I can move the supplier to be only the return type of `allocate` to further reduce how long we keep it. If we want to make that stats calculation eager or not, seems to be unclear. it is a trade-off. I'm ok with both. Considering this balancer calls are one at a time (I think), keeping the last internal state as long as the `allocate` call is in scope doesn't seem that concerning.

Reply:
actually considering `allocate` has a sync and async version, for the `DesiredBalanceShardAllocator` this becomes a bit awkward to return from the listener these stats. I do agree that the extra stats method on the allocator interface is also awkward. Frankly, I think keeping these stats internally in `RoutingNodes` (eagerly calculated) is not a bad option. The pre-calculation doesn't seem that expensive and conceptually not meaningless, because each run of the balancer does produce its own middle state anyway. Also each run of the old balancer does make its changes to the RoutingNodes and records the weights that lead to this decision. So to me it seems recording these stats along with RoutingNodes is the most natural way that fits to the code. Yes, we make multiple calls to the old balancer and we don't use these stats, but that is the overall pattern with how the new balancer wraps the old balancer, e.g. each time the old balancer also rebuilds a model from routing nodes, etc. Any thoughts?

Reply:
also I think what we're trying to export here as metrics, is something very internal to the allocator(s) and escalating it to the interface doesn't have any other use and in a way leaks this internal state/info. So I think `RoutingNodes` and `DesiredBalance` are suitable to pass along this extra bit of info between old allocator, new allocator and the reconciler.

Reply:
For me the discussion began because we wanted to avoid the eager computation since it may be expensive. If that is not true, I don't really have strong opinon on whether it should be part of RoutingNodes or the return value as long as it is not an object field :)

Reply:
I think eager or not is still relevant and orthogonal to whether we add it to the allocator interface or RoutingNodes.

Reply:
going back to eagerly setting the stats in `RoutingNodes` after discussing with Ievgen. 1defb461fb0dc84090f43c79fc0e5fd2923f89bb